src/utils/abi_decoder.ts

```diff
@@ -0,0 +1,66 @@
+import * as Web3 from 'web3';
+import * as _ from 'lodash';
+import {AbiType, DecodedLogArgs, DecodedArgs} from '../types';
+import * as SolidityCoder from 'web3/lib/solidity/coder.js';
+
+export class AbiDecoder {
+    private savedABIs: Web3.AbiDefinition[] = [];
+    private methodIDs: {[signatureHash: string]: Web3.EventAbi} = {};
+    constructor(abiArrays: Web3.AbiDefinition[][]) {
+        _.map(abiArrays, this.addABI.bind(this));
+    }
+    public decodeLog(logItem: Web3.LogEntry): DecodedArgs|undefined {
+        const methodID = logItem.topics[0];
+        const event = this.methodIDs[methodID];
+        if (!_.isUndefined(event)) {
+            const logData = logItem.data;
+            const decodedParams: DecodedLogArgs = {};
+            let dataIndex = 0;
+            let topicsIndex = 1;
+
+            const nonIndexedInputs = _.filter(event.inputs, input => !input.indexed);
+            const dataTypes = _.map(nonIndexedInputs, input => input.type);
+            const decodedData = SolidityCoder.decodeParams(dataTypes, logData.slice(2));
+            _.map(event.inputs, (param: Web3.EventParameter) => {
+                let value;
+                if (param.indexed) {
+                    value = logItem.topics[topicsIndex];
+                    topicsIndex++;
+                } else {
+                    value = decodedData[dataIndex];
+                    dataIndex++;
+                }
+                if (param.type === 'address') {
```

Comment:
Is this an `ABIType`? If so, let's use that type rather then string literal. Same goes for the subsequent ones.

Reply:
It's not an `ABIType`.
```
export enum AbiType {
    Function = 'function',
    Constructor = 'constructor',
    Event = 'event',
    Fallback = 'fallback',
}
```
it's rather `SolidityType`, which we don't have defined yet and I'm rather lazy to define all the `ufixed128x19`s.